### DISCUSSION TITLE
Using bokeh graphics in quarto

### DISCUSSION THREAD

**Discussion by jeremy9959** [quarto-dev/quarto-cli#1744] [Q&A]
Hi,

I prefer using the [bokeh graphics package](https://bokeh.org) but it does not seem to work with quarto.  Has anyone looked into this?
What might be involved in making this possible?

Thanks!

**Comment by cscheid** [comment]
I don't have any further experience with Bokeh, but this worked without fine for me:

    ---
    title: bokeh
    ---
    
    From https://bookdown.org/xiangyun/r-pkgs/chap-reticulate.html
    
    ```{python}
    from bokeh.plotting import figure, output_file, show
    x = [1, 2, 3, 4, 5]
    y = [6, 7, 2, 4, 5]
    output_file("lines.html")
    p = figure(title="simple line example", x_axis_label='x', y_axis_label='y')
    p.line(x, y, legend_label="Temp.", line_width=2)
    show(p)
    ```

**Reply by jeremy9959** [reply]
Sorry if I wasn't clear -- rendering to html does work, but not to pdf.

For example:

---
title: "Bokeh Plot"
format: html
---

```{python}
from bokeh.plotting import figure, output_file, show, output_notebook
x = [1, 2, 3, 4, 5]
y = [6, 7, 2, 4, 5]
output_notebook()
p = figure(title="simple line example", x_axis_label='x', y_axis_label='y')
p.line(x, y, legend_label="Temp.", line_width=2)
show(p)
```

works fine, but if you specify ```pdf``` as the format you get:
![image](https://user-images.githubusercontent.com/11709993/183477146-b9365471-2eda-43d8-835d-1c9624b923d2.png)

**Reply by cscheid** [reply]
The issue is that those are interactive plots, and there's no general support for for interactivity in PDF.

You're going to have to get bokeh to emit a PNG or SVG image, and I don't know enough about it to help there.

**Reply by jeremy9959** [reply]
Somehow this is done for plotly plots, I assume?  You can embed the interactive plotly graphics in html, but they
get static rendering for pdf?  I don't use plotly so I am not sure how it works.

**Reply by jtrecenti** [reply]
I think `{knitr}` package from R uses [webshot](https://github.com/wch/webshot) or [webshot2](https://github.com/rstudio/webshot2), which takes a screenshot of the visualization to get the PNG. More information [here](https://bookdown.org/yihui/bookdown/html-widgets.html#html-widgets). Maybe there is an equivalent to that in python to render the image?

**Reply by jeremy9959** [reply]
That's a good hint! I will look into it.  Using a screenshot costs a lot of resolution, but it's worth a try.

**Reply by jeremy9959** [reply]
It seems I was naive.  I can't get plotly plots to embed in pdf either.  I think I will close this discussion and open an issue raising the question generally.
